Adds DataLogger class

LaPommeCosmique opened the pull request:
Creates the DataLogger class, with functionality to create, open, and modify csv files on the SD card.

Specifications on the the DataLogger class can be found in [this comment](https://github.com/supermileage/dynamometer-firmware/issues/15#issuecomment-1472659332) in issue #15

underscoreHasan committed:
Adds code to render force and angular velocity readings on display
Debugging code left in but commented out

underscoreHasan committed:
initial commit
adds code to render force and angular velocity on display

underscoreHasan committed:
adds code to initialize SD Card and read .txt file from it

underscoreHasan committed:
begins building framework for saving csv files
sets baud rate in platformio.ini

underscoreHasan committed:
incomplete framework for writing csv files
TODO: encapsulate csv writing behavior in a seperate class

underscoreHasan committed:
First version of DataLogger class
TODO: complete DataLogger::loadCSV()
test and debug class

underscoreHasan committed:
Completes DataLogger::loadCSV
adds debugging print statements
TODO: testing through main.cpp

underscoreHasan committed:
fixes bug in DataLogger::loadCSV()

Commit:
fixed boolean equalities and string concatenation

LaPommeCosmique committed:
fixed boolean equalities and string concatenation

Commit:
refactored function and parameter names to match specs

LaPommeCosmique committed:
refactored function and parameter names to match specs

Commit:
updated access of setHeader and addRow from private to public

LaPommeCosmique committed:
updated access of setHeader and addRow from private to public

Commit:
removed writeLog function as not present in specs

LaPommeCosmique committed:
removed writeLog function as not present in specs

Commit:
added saveToDisk, addEntry, and getBufferLength. Implemented string buffer instead of immediately writing to files

LaPommeCosmique committed:
added saveToDisk, addEntry, and getBufferLength. Implemented string buffer instead of immediately writing to files

Commit:
added O_SYNC flag

LaPommeCosmique committed:
added O_SYNC flag

Commit:
removed <vector> library from DataLogger and added missing semicolor

LaPommeCosmique committed:
removed <vector> library from DataLogger and added missing semicolor

Commit:
updated algorithm to count the number of columns when opening a csv file

LaPommeCosmique committed:
updated algorithm to count the number of columns when opening a csv file

LaPommeCosmique committed:
removed infinite loop upon missing SD card

Commit:
removed infinite loop upon missing SD card

LaPommeCosmique committed:
updated open() functionality to match flow diagram in issue #15

Commit:
updated open() functionality to match flow diagram in issue #15

Commit:
Merge branch 'dev_sdcard' of github.com:supermileage/dynamometer-firmware into dev_sdcard

LaPommeCosmique committed:
Fix compiler errors

LaPommeCosmique committed:
Revise algorithm for generating file names

LaPommeCosmique committed:
Merge branch 'dev' into dev_sdcard

LaPommeCosmique committed:
Change null check of current file for consistency

LaPommeCosmique committed:
Merge branch 'dev_sdcard' of github.com:supermileage/dynamometer-firmware into dev_sdcard

LaPommeCosmique commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+        (*_curFilePtr).close();
+
+        // sets _curFile to null so saveToDisk() doesnt try writing to a closed file
+        _curFilePtr = nullptr;
```
I changed ```_curFile``` to ```_curFilePtr``` because objects cannot be set to null, while pointers can.

@cosparks do we have to run something like ```delete (*_curFilePtr)``` to prevent memory leaks?

cosparks replied:
You only need to use the ```delete``` operator if you've used the ```new``` operator somewhere.  In general, you only need to worry about memory leaks if you're calling ```new``` or any of the C dynamic memory allocation functions (```malloc```, ```alloc``` etc..)

LaPommeCosmique replied:
Ah okay thats helpful, thanks

cosparks commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+        {
+            // file is empty
+            _numColumns = numColumns;
+            _curFilePtr = &tempFile;
```
Using pointers is a bit tricky.  ```tempFile``` is on the stack and its lifetime is limited to the scope of this call to ```DataLogger::open```.  This means that memory at the address ```&tempFile``` will eventually be overwritten (although we don't know when) and ```_curFilePtr``` will become a dangling pointer, which essentially means that you have a valid address which points invalid (garbage) data.

There are a number of ways of dealing with this, the easiest of which would be to not use a pointer reference here, and instead copy the value of ```tempFile``` to a non-pointer ```_curFile```.  I would go even further and say that you should just assign the return from SD.open directly to ```_curFile``` (```_curFile = SD.open(name)```).  If a user has called DataLogger::open, they probably don't care about the previously open file, so we can just overwrite it right away.

cosparks replied:
DataLogger::create is a simpler method and has the same problem, so is maybe a clearer example of what I mean

LaPommeCosmique replied:
Ya pointers are a headache.. I replaced it with _curFile as you suggested

cosparks replied:
ok so apparently the File class overloads the bool() operator so you can treat it as a nullable object:

```
if (_curFile) {
  // valid
} else {
  // invalid
}
```

This is kind of strange, but but it means that you don't need to maintain that bool.

LaPommeCosmique replied:
If we toss the ```_fileValid``` variable, the only thing I am concerned about is if ```_curFile``` still evaluates to true after we run ```_curFile.close()```. This is relevant when we are writing to the file, when check whether the file handle is still valid. 

Do you know the answer to this? I wish there was better documentation about this online, as I am not able to find anything conclusive.

cosparks replied:
Does it evaluate to true after you call close?  You can try for yourself:

```
// after initializing everything in setup():
File myFile = SD.open("somePath", FILE_READ);
if (myFile)
  Serial.println("File opened!");
else
  Serial.println("error opening file");

myFile.close();

if (myFile)
  Serial.println("bool() operator evaluates to true after call to close");
else
  Serial.println("bool() operator evaluates to false after call to close");
 ```

LaPommeCosmique replied:
I tested it and it seems like it still evaluates to true after we call ```myFile.close()```. Its kinda strange.. but it seems like we would need to keep track of the file status with the ```_fileValid``` variable.

cosparks replied:
Does it evaluate to true for my mock API or for the real arduino file API?

I might not have implemented it properly in my arduino filesystem mocks

cosparks replied:
@LaPommeCosmique I just fixed this so the mock file should evaluate to false when it is not valid

cosparks replied:
I just checked the Arduino File implementation, and when you call close it does set its internal file pointer to null, so you should be able to depend on this behavior (and so the _fileValid variable is not needed)

cosparks replied:
Done.  I'll resolve this discussion

cosparks commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+//public methods:
+
+DataLogger::DataLogger(int pinNumber, bool O_SYNC_FLAG /*= false*/)
+{
```
For style, please refer to elec [c++ style guide](https://www.notion.so/C-Style-Guide-f40ef59a0c3140148e1e578bfc61f1b8):

cosparks replied:
Sorry, that probably wasn't clear.  In addition to naming conventions, opening curly braces should be at the end of the line above, not on their own line.

It's a small thing, but something that is pretty much consistent across all of elec's codebases

LaPommeCosmique replied:
Fix has been implemented

cosparks commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+
+//public methods:
+
+DataLogger::DataLogger(int pinNumber, bool O_SYNC_FLAG /*= false*/)
```
only use SCREAMING_SNAKE_CASE for macros.  Param name can just be ```sync``` or ```osync```.

cosparks commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+bool DataLogger::create(String name, int numColumns)
+{
+    File curFile = SD.open(name, FILE_WRITE);
+    _curFilePtr = &curFile;
```
same issue with pointers applies here.  When this method returns, the address of ```curFile``` is no longer valid.

cosparks commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+    _curColumn = 0;
+    _buffer = "";
+    
+    if (_curFilePtr != nullptr)
```
Just fyi, ```_curFilePtr``` will never be null here because you always assign it a valid stack address above.

cosparks commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+            {
+                commaNum++;
+            }
+            else if (readChar == '\n' || readChar == '\r')
```
You could just check for ```'\n'``` here, which would catch both ```LF``` and ```CRLF```

LaPommeCosmique replied:
I do agree, but, this made me think..

If there was a file that was only a CRLF
```
\r\n
```
then only checking for ```'\n'``` would cause the file to be interpreted as having one column, even though it arguably would be a csv file with zero columns. Because of this, in my opinion, including the ```'\r'``` would make the code more thorough.

cosparks replied:
That's fine.  In general I would trust that the csv file is formatted as defined [here](https://www.rfc-editor.org/rfc/rfc4180), so I wouldn't worry too much about edge cases like that.

cosparks commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+            String extension = extensionIndex == -1 ? "" : name.substring(extensionIndex + 1);
+            base.trim();
+
+            // find current file count (example: myFile (2).csv)
```
does this code add a space in between file name and (#)?

LaPommeCosmique replied:
Yup, it does in line 152. Do you want this to be the intended behaviour?

cosparks replied:
Spaces in paths are a pain.  I think removing the space is a good idea

LaPommeCosmique replied:
Space has been removed

cosparks commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+        else 
+        {
+            // file is not empty, and does not have the correct number of columns
+            DEBUG_SERIAL_LN("Incorrect column number. Creating in file copyof_" + name);
```
maybe change this to "Creating in fileName(n).csv"

LaPommeCosmique replied:
I updated the output to
```
DEBUG_SERIAL_LN("Incorrect column number in file: " + name + ". Creating file with new name.");
```
as we still have to parse the file name. 

Then, once we call ```open()``` with the new file name it will print
```
DEBUG_SERIAL_LN("Loading " + *newFileName*);
```

cosparks commented on src/DataLogger/DataLogger.h:
```diff
@@ -0,0 +1,58 @@
+class DataLogger {
+    public:
+
+        //Constructs a DataLogger object and begins access to the SDcard.
```
pls format comments as in other files in this project:

```
/**
 * @brief (description here)
 * ...
*/
```

This formatting offers many options for adding documentation on return types/params/etc.. and will format it all for you.  Try going to main.cpp and hovering over one of these method calls to see what I mean:

```
tft.begin();
tft.setRotation(3);
tft.fillScreen(COLOUR_BLACK);
```

These aren't from our source code, but they do use the same style of documentation

cosparks commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+}
+
+bool DataLogger::saveToDisk() {
+    if (_curFilePtr != nullptr)
```
Considering the changes I've recommended above, if you want to check that there is a currently open valid file before you go to save, you can maintain a bool value (```DataLogger::_fileValid```) which indicates whether or not ```DataLogger::_curFile``` is valid

LaPommeCosmique replied:
Good idea, I have implemented this change

cosparks replied:
Copying the same comment from above:

> Apparently the File class overloads the bool() operator so you can treat it as a nullable object:

```
if (_curFile) {
  // valid
} else {
  // invalid
}
```

> This is kind of strange, but but it means that you don't need to maintain that bool.

cosparks commented on src/DataLogger/DataLogger.cpp:
```diff
@@ -0,0 +1,248 @@
+
+bool DataLogger::close()
+{
+    if (_curFilePtr != nullptr)
```
Same here, instead of using a pointer you can maintain a bool ```DataLogger::_fileValid```

cosparks commented:
Good work!  There is some pretty tricky logic in here.

I think it might be worth writing some unit tests that we can run and debug natively so we can verify that everything works as expected.  I'll work on adding mock implementations of the arduino filesystem.

LaPommeCosmique committed:
Add documentation and fix style issues

LaPommeCosmique committed:
Replace file pointer with file object and revise file creation system

LaPommeCosmique commented:
Thanks! I think having the ability to test the code on a mock filesystem would be helpful. 

I tried uploading the code to my pico board to run some testing (specifically the testing_sdcard branch), but I am having trouble consistently reading serial messages. Or more accurately - I received one message over serial and I wasnt able to read any messages anymore. I am trying to figure out the issue, and whether its a problem with the code or my laptop.

cosparks commented:
> ... Or more accurately - I received one message over serial and I wasnt able to read any messages anymore. I am trying to figure out the issue, and whether its a problem with the code or my laptop.

This could be many things, but hopefully something we could catch with a test.  I started working on a simple mock file system which should be done by tomorrow

LaPommeCosmique committed:
Fix styling inconsistencies

LaPommeCosmique committed:
Remove space in file name format

cosparks committed:
Add tests for DataLogger (#41)

* creates implementation of mock filesystem
* adds two example tests - one fails
* moves DataLogger to System folder
* adds means of generating input test data programmatically

cosparks committed:
merges recent changes from dev

cosparks committed:
Fixes .gitignore

cosparks commented on src/System/DataLogger.cpp:
```diff
@@ -0,0 +1,232 @@
+//public methods:
+
+DataLogger::DataLogger(int pinNumber, bool osync /*= false*/) {
+    if (SD.begin(pinNumber, SPI1)) {
```
I just noticed this.  All system begin / init methods should be called from setup.

DataLogger is going to be a global object, so its constructor will be called during this program's global initialization.  Generally, you want to make sure that all system resources are initialized before you make any system calls, so you should move this code to an init method which we can call from the main setup function.

cosparks committed:
Modifications to suppress sign-compare warnings

LaPommeCosmique committed:
Move system calls to init method

LaPommeCosmique committed:
Update initialization of DataLogger object in tests

cosparks committed:
Fixes bug in mock File::available implementation

LaPommeCosmique committed:
Modify tests

LaPommeCosmique committed:
Merge pull request #56 from supermileage/dev-file-fix

Fixes bug in mock File::available implementation

LaPommeCosmique committed:
Fix entry adding bug

LaPommeCosmique committed:
Add column check for add row

LaPommeCosmique committed:
Add getFileName function to get file name

LaPommeCosmique committed:
Fix _curFile overwriting bug

LaPommeCosmique committed:
Update naming functionality for open and create functions

LaPommeCosmique committed:
Update DataLogger function briefs

LaPommeCosmique committed:
Change open() function to read until end of file so append occurs

LaPommeCosmique committed:
Add datalogger unit tests

LaPommeCosmique commented:
I added a few more changes and added some unit tests. I believe all the previous issues have been hammered out with this latest push. 

Here is a summary of the changes to the original ```DataLogger``` specs
- The ```init()``` function must be called when initializing DataLogger
- The ```getFileName()``` function returns the file name of the currently open file (or a blank string if no file is open)
- The ```open()``` function creates a file with the specified name only if the file doesn't exist - otherwise, it generates a new name

cosparks committed:
Modifies SD::File mock to evaluate to false if file has been closed

cosparks committed:
Removes _fileValid member from DataLogger
